// react-timer-wrapper/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _cleanReactProps = _interopRequireDefault(require("clean-react-props"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

var TimerWrapper =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(TimerWrapper, _Component);

  function TimerWrapper(props) {
    var _this;

    _this = _Component.call(this, props) || this;
    _this.state = {
      duration: props.duration,
      startTime: props.active ? Date.now() - props.time : null,
      time: props.time
    };
    _this.tick = _this.tick.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  var _proto = TimerWrapper.prototype;

  _proto.componentDidMount = function componentDidMount() {
    var _this$state = this.state,
        duration = _this$state.duration,
        startTime = _this$state.startTime,
        time = _this$state.time;
    var _this$props = this.props,
        active = _this$props.active,
        onStart = _this$props.onStart;

    if (active) {
      onStart({
        duration: duration,
        progress: this.getProgress(time),
        time: time
      });
      this.animationFrame = requestAnimationFrame(this.tick);
    }
  };

  _proto.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var _this2 = this;

    var active = nextProps.active,
        duration = nextProps.duration,
        time = nextProps.time,
        onStart = nextProps.onStart,
        onStop = nextProps.onStop;

    if (active === this.props.active && time !== this.props.time && time !== this.state.time) {
      var timeDiff = this.state.time - time;
      this.setState({
        startTime: this.state.startTime + timeDiff,
        time: this.state.time + timeDiff
      });
      return;
    }

    if (active !== this.props.active) {
      switch (active) {
        case true:
          var nextTime = duration !== null && this.state.time >= duration ? 0 : this.state.time;
          this.setState({
            startTime: Date.now() - nextTime,
            time: nextTime
          }, function () {
            onStart({
              duration: duration,
              progress: _this2.getProgress(nextTime),
              time: nextTime
            });
            _this2.animationFrame = requestAnimationFrame(_this2.tick);
          });
          break;

        case false:
          cancelAnimationFrame(this.animationFrame);
          onStop({
            duration: duration,
            progress: this.getProgress(this.state.time),
            time: this.state.time
          });
          break;
      }
    }
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    cancelAnimationFrame(this.animationFrame);
  };

  _proto.getProgress = function getProgress(time) {
    var duration = this.state.duration;

    if (!duration) {
      return 0;
    }

    return Math.max(0, Math.min(1, time / duration));
  };

  _proto.tick = function tick() {
    var _this3 = this;

    var _this$props2 = this.props,
        duration = _this$props2.duration,
        loop = _this$props2.loop,
        time = _this$props2.time,
        onFinish = _this$props2.onFinish,
        onTimeUpdate = _this$props2.onTimeUpdate,
        onStart = _this$props2.onStart;
    var startTime = this.state.startTime;
    var nextTime = Date.now() - startTime;
    this.setState({
      time: nextTime
    }, function () {
      onTimeUpdate({
        duration: duration,
        progress: _this3.getProgress(nextTime),
        time: nextTime
      });

      if (duration !== null && nextTime >= duration) {
        onFinish({
          duration: duration,
          progress: _this3.getProgress(nextTime),
          time: nextTime
        });

        if (!loop) {
          cancelAnimationFrame(_this3.animationFrame);
          return;
        }

        nextTime = 0;
        onStart({
          duration: duration,
          progress: 0,
          time: nextTime
        });

        _this3.setState({
          startTime: Date.now()
        });
      }

      _this3.animationFrame = requestAnimationFrame(_this3.tick);
    });
  };

  _proto.render = function render() {
    var _this$props3 = this.props,
        children = _this$props3.children,
        component = _this$props3.component,
        duration = _this$props3.duration;
    var time = this.state.time;
    var progress = duration && time ? time / duration : 0;

    var clonedChildren = _react.default.Children.toArray(children).map(function (child, childIndex) {
      return _react.default.cloneElement(child, {
        duration: duration,
        time: time,
        progress: progress
      });
    });

    return _react.default.createElement(component, _objectSpread({}, (0, _cleanReactProps.default)(this.props)), clonedChildren);
  };

  return TimerWrapper;
}(_react.Component);

TimerWrapper.propTypes = {
  active: _propTypes.default.bool,
  component: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  duration: _propTypes.default.number,
  loop: _propTypes.default.bool,
  time: _propTypes.default.number,
  onFinish: _propTypes.default.func,
  onTimeUpdate: _propTypes.default.func,
  onStart: _propTypes.default.func
};
TimerWrapper.defaultProps = {
  active: false,
  component: 'div',
  duration: 10000,
  loop: false,
  time: 0,
  onFinish: function onFinish() {},
  onStart: function onStart() {},
  onStop: function onStop() {},
  onTimeUpdate: function onTimeUpdate() {}
};
var _default = TimerWrapper;
exports.default = _default;
